data_structures/rtree.cpp

```diff
@@ -0,0 +1,168 @@
+// rtree.cpp
+
+#include <iostream>
+#include <memory>
+
+#include "rtree.hpp"
+
+using coord_t = spatial::coord_t;
+using code_t = spatial::code_t;
+using index_t = spatial::index_t;
+
+int const M = 8;
+
+template<typename T>
+spatial::Rtree<T>::Rtree(): 
+    root_entry(std::make_shared<InternalEntry>()) 
+{
+    root_entry->bounding_box = (Rectangle){0, 0, 0, 0};
+    root_entry->node = std::make_shared<Node>(*this);
+}
+
+// ~Rtree() {}
+
+template<typename T>
+spatial::Rtree<T>::Node::Node(Rtree& rt): 
+    load(0), 
+    m(0), 
+    rtree(rt) 
+{ }
+
+// ~Node() {}
+
+/**
+ * Construct an R-tree from the given point data.
+ * Currently, this is just doing point-by-point insertion
+ */
+template<typename T>
+void spatial::Rtree<T>::build(std::vector<T> const raw_data) {
+    // Pick an inital bounding box for the root
+    Rectangle initial_seed = {
+        raw_data[0][0], raw_data[0][0],
+        raw_data[0][1], raw_data[0][1]
+    };
+    root_entry->bounding_box = initial_seed;
+
+    // Insert each data point into the R-tree
+    data.reserve(raw_data.size());
+    for (auto const& raw_datum : raw_data) {
+        Point const new_point = {raw_datum[0], raw_datum[1]};
+        Datum<T> const new_datum = {raw_datum, new_point};
+        insert(new_datum);
+    } 
+}
+
+template<typename T>
+void spatial::Rtree<T>::insert(Datum<T> new_datum) {
+    data.push_back(new_datum);
+
+    // Expand the root's bounding box if necessary
+    root_entry->bounding_box = min_bounding_box(
+        root_entry->bounding_box, new_datum.point
+    );
+
+    // Recursively insert the point into the root node
+    if (root_entry->node->insert(data.size()-1)) {
+        split_root();
+    }
+}
+
+/**
+ * Recursively insert a point into the current node.
+ * If a node exceeds M entries, we return 'true' to indicate that a split
+ * is required, since splitting happens at the parent's level.  
+ */
+template<typename T>
+bool spatial::Rtree<T>::Node::insert(index_t const point_idx) {
+    Point const p = rtree.data[point_idx].point;
+    if (this->is_leaf()) {
+        // add the point to the current node
+        auto new_entry = std::make_shared<LeafEntry>();
+        new_entry->bounding_box = (Rectangle){ p.x, p.x, p.y, p.y };
+        new_entry->idx = point_idx;
+        entries.push_back(new_entry);
```

Comment:
```suggestion
        entries.push_back( std::make_unique<LeafEntry>{(Rectangle){ p.x, p.x, p.y, p.y }
                                                      , point_idx});
```
Important difference here is the shift to a _unique_ pointer, rather than shared. So far as I can tell, you don't have separate objects that share ownership of this entry. (Note also that this change cascades to your `.hpp` file.)

Reply:
Is there an advantage to using a `unique_ptr` rather than a `shared_ptr`? From what I've read, it seems like `shared_ptr` is just more general, in case I do come across a situation where I'd want a second pointer to an object.

Reply:
Who wants to argue with the [ISO C++ Core Guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-owner)? :smile:

A shared pointer has overhead in order to count references, which is reason enough for me. But the choice is also _self-documenting_: when you use a `std::unique_ptr` it is clear that this object belongs here; when you use a `std::shared_ptr`, one immediately starts wondering "who else owns this object and what is its lifespan"?

To stretch your analogy, a `std::set<>` is more general than a `std::unordered_set<>`. But I would always default to the latter and only use an _ordered_ set if I encountered a situation in which I needed the order, because the `std::unordered_set<>` is almost guaranteed to be faster with no extra effort on my part. There is no free lunch, and generality typically comes with a cost.

Reply:
Okay, that's a good point, thank you. Also, far be it from me to argue with the ISO!